File: ios/QuentinoApp/Services/Stockin.swift
import Foundation

/**
 Naskladnění — naskladnění zboží.

 Zboží dorazí v krabici, někdo ho projde kus po kuse a v e-shopu musí přibýt
 na skladě. Dělá se to u regálu, ne u počítače, takže se to **začíná na
 telefonu** a dokončuje na Macu — a mezi tím se nesmí nic ztratit.

 Naskladnění je proto seznam řádků, ne jeden dokument: každý řádek je kód
 a počet. Řádky se mezi zařízeními slučují po jednom (stejně jako deníky
 poukazů), takže když se na telefonu přidá pět položek a na Macu tři,
 výsledek má osm — bez ohledu na to, které zařízení bylo první.

 Zápis do e-shopu tady není schválně: okno administrace se dá otevřít jen na
 počítači. Telefon naskladnění připraví a pošle ji sdílenou složkou dál.
 */
enum Stockin {
    private static func touch(_ id: String) {
        try? SQLite.shared.run("UPDATE stockin SET updated_at = ? WHERE id = ?",
                               [.text(Formats.iso()), .text(id)])
    }

    private static func shape(_ row: [String: Any]) -> [String: Any] {
        [
            "id": row["id"] ?? "",
            "title": row["title"] ?? "",
            "note": row["note"] ?? "",
            "device": row["device"] ?? "",
            "state": (row["state"] as? String) == "sent" ? "sent" : "open",
            "createdAt": row["created_at"] ?? "",
            "updatedAt": row["updated_at"] ?? "",
            "sentAt": (row["sent_at"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? NSNull(),
            "lines": row["lines"] ?? 0,
            "pieces": row["pieces"] ?? 0
        ]
    }

    private static let counted = """
    SELECT s.*, (SELECT COUNT(*) FROM stockin_items i WHERE i.session_id = s.id) AS lines,
           (SELECT COALESCE(SUM(qty), 0) FROM stockin_items i WHERE i.session_id = s.id) AS pieces
    FROM stockin s
    """

    static func list() -> [[String: Any]] {
        let rows = (try? SQLite.shared.query(
            counted + " WHERE s.state != 'deleted' ORDER BY (s.state = 'open') DESC, s.updated_at DESC LIMIT 60"
        )) ?? []
        return rows.map(shape)
    }

    static func session(_ id: String) -> Any {
        guard let row = (try? SQLite.shared.query(
            counted + " WHERE s.id = ? AND s.state != 'deleted'", [.text(id)]
        ))?.first else { return NSNull() }
        return shape(row)
    }

    static func create(_ title: String) -> Any {
        let id = "\(Device.id().prefix(6))-\(String(Int(Date().timeIntervalSince1970 * 1000), radix: 36))"
        let at = Formats.iso()
        let stamp = DateFormatter()
        stamp.locale = Locale(identifier: "cs_CZ")
        stamp.dateStyle = .medium
        stamp.timeStyle = .none
        let name = title.isEmpty ? "Naskladnění \(stamp.string(from: Date()))" : title
        try? SQLite.shared.run(
            "INSERT INTO stockin (id, title, device, state, created_at, updated_at) VALUES (?,?,?,'open',?,?)",
            [.text(id), .text(name), .text(Device.name()), .text(at), .text(at)]
        )
        return session(id)
    }

    static func items(_ id: String) -> [[String: Any]] {
        let rows = (try? SQLite.shared.query(
            "SELECT * FROM stockin_items WHERE session_id = ? ORDER BY added_at DESC", [.text(id)]
        )) ?? []
        return rows.map { row in
            [
                "code": row["code"] ?? "",
                "productCode": row["product_code"] ?? "",
                "title": row["title"] ?? "",
                "label": row["label"] ?? "",
                "qty": row["qty"] ?? 0,
                "stockBefore": row["stock_before"] ?? NSNull(),
                "addedAt": row["added_at"] ?? ""
            ]
        }
    }

    /**
     Přidá načtený kód. Když už na naskladnění je, jen se přičte počet.

     Když se nenajde nic, **nic se nepřidá**: naskladnění s řádkem „neznámý kód"
     by se nedala odeslat a při zápisu do e-shopu by se stejně musela řešit
     ručně.
     */
    static func scan(_ id: String, _ raw: String, qty: Int) -> [String: Any] {
        guard let found = Catalog.find(raw) as? [String: Any],
              let code = found["code"] as? String else {
            return ["added": false, "unknown": raw.trimmingCharacters(in: .whitespacesAndNewlines)]
        }
        let count = max(1, qty)
        let existing = (try? SQLite.shared.query(
            "SELECT qty FROM stockin_items WHERE session_id = ? AND code = ?", [.text(id), .text(code)]
        ))?.first

        if existing != nil {
            try? SQLite.shared.run(
                "UPDATE stockin_items SET qty = qty + ? WHERE session_id = ? AND code = ?",
                [.int(Int64(count)), .text(id), .text(code)]
            )
        } else {
            let stock = found["stock"] as? Int
            try? SQLite.shared.run(
                """
                INSERT INTO stockin_items (session_id, code, product_code, title, label, qty, stock_before, added_at)
                VALUES (?,?,?,?,?,?,?,?)
                """,
                [
                    .text(id), .text(code),
                    .text(found["productCode"] as? String ?? ""),
                    .text(found["title"] as? String ?? ""),
                    .text(found["label"] as? String ?? ""),
                    .int(Int64(count)),
                    stock.map { SQLite.Value.int(Int64($0)) } ?? .null,
                    .text(Formats.iso())
                ]
            )
        }
        touch(id)
        let item = items(id).first { ($0["code"] as? String) == code }
        return ["added": true, "item": item ?? NSNull()]
    }

    static func setQty(_ id: String, _ code: String, _ qty: Int) {
        if qty <= 0 {
            try? SQLite.shared.run("DELETE FROM stockin_items WHERE session_id = ? AND code = ?",
                                   [.text(id), .text(code)])
        } else {
            try? SQLite.shared.run("UPDATE stockin_items SET qty = ? WHERE session_id = ? AND code = ?",
                                   [.int(Int64(qty)), .text(id), .text(code)])
        }
        touch(id)
    }

    static func rename(_ id: String, title: String, note: String) {
        try? SQLite.shared.run("UPDATE stockin SET title = ?, note = ?, updated_at = ? WHERE id = ?",
                               [.text(title), .text(note), .text(Formats.iso()), .text(id)])
    }

    /**
     Smazání se musí dozvědět i druhé zařízení.

     Řádek se proto nezahodí, jen se označí za smazaný a takový zůstane
     ve sdíleném souboru. Kdyby se smazal doopravdy, počítač by ho při
     nejbližší synchronizaci poslal zpátky — přesně to se dělo: co se smazalo
     na telefonu, do minuty se zase objevilo. Náhrobky se po dvou měsících
     uklidí, to už se nemá co vracet.
     */
    static func remove(_ id: String) {
        try? SQLite.shared.run("DELETE FROM stockin_items WHERE session_id = ?", [.text(id)])
        try? SQLite.shared.run("UPDATE stockin SET state = 'deleted', updated_at = ? WHERE id = ?",
                               [.text(Formats.iso()), .text(id)])
        let cutoff = Formats.iso(Date().addingTimeInterval(-60 * 86_400))
        try? SQLite.shared.run("DELETE FROM stockin WHERE state = 'deleted' AND updated_at < ?",
                               [.text(cutoff)])
    }

    static func markSent(_ id: String) {
        let at = Formats.iso()
        try? SQLite.shared.run("UPDATE stockin SET state = 'sent', sent_at = ?, updated_at = ? WHERE id = ?",
                               [.text(at), .text(at), .text(id)])
    }

    /**
     Podklad pro zápis do e-shopu.

     Vnitřní čísla (`product_id`, `variant_id`) se doplní z feedu — bez nich
     Upgates zápis nepřijme. Co se dohledat nepodaří, se nezamlčí: vrátí se
     s prázdným `productId` a rozhraní to ukáže jako řádek k ručnímu dořešení.
     */
    static func plan(_ id: String) -> [[String: Any]] {
        items(id).map { item in
            let code = item["code"] as? String ?? ""
            let variant = (try? SQLite.shared.query(
                "SELECT * FROM product_variants WHERE code = ?", [.text(code)]
            ))?.first
            let productCode = (variant?["product_code"] as? String) ?? code
            let product = (try? SQLite.shared.query(
                "SELECT * FROM products WHERE code = ?", [.text(productCode)]
            ))?.first
            let stockNow = (variant?["stock"] as? Int) ?? (product?["stock"] as? Int)
            let before = item["stockBefore"] as? Int

            return [
                "code": code,
                "title": (item["title"] as? String).flatMap { $0.isEmpty ? nil : $0 }
                    ?? (product?["title_cz"] as? String) ?? code,
                "label": item["label"] ?? "",
                "qty": item["qty"] ?? 0,
                "productId": product?["product_id"] ?? "",
                "variantId": variant?["variant_id"] ?? "",
                "stockNow": stockNow.map { $0 as Any } ?? NSNull(),
                "stockBefore": before.map { $0 as Any } ?? NSNull(),
                // Zásoba se od načtení změnila — mezitím se prodalo nebo někdo
                // naskladnil ručně. Není to chyba, ale je to jediná informace,
                // kvůli které se má člověk před odesláním podívat.
                "moved": before != nil && stockNow != nil && before != stockNow
            ]
        }
    }

    // MARK: - Synchronizace mezi zařízeními

    /**
     Sloučení naskladnění ze sdílené složky.

     **Vyhrává novější verze celého naskladnění, ne jednotlivé řádky.**

     Původně se slučovalo po řádcích a počet se bral jako vyšší z obou stran,
     aby se opakovanou synchronizací nic nenafouklo. Jenže tím se nedalo počet
     snížit ani řádek smazat: soubor z minulého kola hodnotu vždycky vrátil.
     Projevilo se to i na jediném zařízení — opravený počet se do minuty vrátil
     na původní, protože ho vzkřísil soubor, který to samé zařízení předtím
     zapsalo.

     Naskladnění se dělá po sobě, ne najednou: začne se u regálu na telefonu
     a dokončí na počítači. Novější strana proto přebírá i seznam řádků —
     včetně toho, co z něj zmizelo.

     Dvě věci jsou konečné bez ohledu na čas: **smazané** a **zapsané**.
     */
    static func merge(_ remote: [String: Any]) {
        let sessions = remote["sessions"] as? [[String: Any]] ?? []

        /*
         Nejdřív se zjistí, čí verze je novější — porovnává se se stavem
         *před* zápisem hlavičky, protože ta si po sloučení odnese pozdější
         čas z obou stran a rozdíl by zmizel.
         */
        var takeover = Set<String>()
        for row in sessions {
            guard let id = row["id"] as? String, !id.isEmpty else { continue }
            let mine = (try? SQLite.shared.query(
                "SELECT updated_at, state FROM stockin WHERE id = ?", [.text(id)]
            ))?.first
            if (mine?["state"] as? String) == "deleted" { continue }
            let theirs = row["updated_at"] as? String ?? ""
            if mine == nil || theirs > (mine?["updated_at"] as? String ?? "") { takeover.insert(id) }
        }

        for row in sessions {
            guard let id = row["id"] as? String, !id.isEmpty else { continue }
            let state = row["state"] as? String ?? "open"
            try? SQLite.shared.run(
                """
                INSERT INTO stockin (id, title, note, device, state, created_at, updated_at, sent_at)
                VALUES (?,?,?,?,?,?,?,?)
                ON CONFLICT(id) DO UPDATE SET
                  title = CASE WHEN excluded.updated_at > stockin.updated_at THEN excluded.title ELSE stockin.title END,
                  note = CASE WHEN excluded.updated_at > stockin.updated_at THEN excluded.note ELSE stockin.note END,
                  state = CASE
                    WHEN stockin.state = 'deleted' OR excluded.state = 'deleted' THEN 'deleted'
                    WHEN excluded.state = 'sent' THEN 'sent'
                    ELSE stockin.state END,
                  sent_at = CASE WHEN stockin.sent_at = '' THEN excluded.sent_at ELSE stockin.sent_at END,
                  updated_at = MAX(stockin.updated_at, excluded.updated_at)
                """,
                [
                    .text(id),
                    .text(row["title"] as? String ?? ""),
                    .text(row["note"] as? String ?? ""),
                    .text(row["device"] as? String ?? ""),
                    .text(["sent", "deleted"].contains(state) ? state : "open"),
                    .text(row["created_at"] as? String ?? Formats.iso()),
                    .text(row["updated_at"] as? String ?? Formats.iso()),
                    .text(row["sent_at"] as? String ?? "")
                ]
            )
        }

        // Přebírané naskladnění dostane cizí seznam řádků celý — a to znamená
        // i to, že řádky, které v něm nejsou, tady zmizí
        for id in takeover {
            try? SQLite.shared.run("DELETE FROM stockin_items WHERE session_id = ?", [.text(id)])
        }

        for row in remote["items"] as? [[String: Any]] ?? [] {
            guard let session = row["session_id"] as? String, takeover.contains(session),
                  let code = row["code"] as? String, !code.isEmpty else { continue }
            let stock = row["stock_before"] as? Int
            try? SQLite.shared.run(
                """
                INSERT OR REPLACE INTO stockin_items
                  (session_id, code, product_code, title, label, qty, stock_before, added_at)
                VALUES (?,?,?,?,?,?,?,?)
                """,
                [
                    .text(session), .text(code),
                    .text(row["product_code"] as? String ?? ""),
                    .text(row["title"] as? String ?? ""),
                    .text(row["label"] as? String ?? ""),
                    .int(Int64(row["qty"] as? Int ?? 0)),
                    stock.map { SQLite.Value.int(Int64($0)) } ?? .null,
                    .text(row["added_at"] as? String ?? Formats.iso())
                ]
            )
        }

        // Řádky smazaného naskladnění se nevracejí — ani z druhé strany
        try? SQLite.shared.run(
            "DELETE FROM stockin_items WHERE session_id IN (SELECT id FROM stockin WHERE state = 'deleted')"
        )
    }

    /// Co se má poslat do sdílené složky — rozpracované, nedávno odeslané
    /// a **náhrobky po smazaných**, jinak by se smazané vracelo zpátky.
    static func export() -> [String: Any] {
        let cutoff = Formats.iso(Date().addingTimeInterval(-30 * 86_400))
        let sessions = (try? SQLite.shared.query(
            "SELECT * FROM stockin WHERE state = 'open' OR updated_at > ? ORDER BY updated_at DESC LIMIT 120",
            [.text(cutoff)]
        )) ?? []
        let ids = sessions.compactMap { $0["id"] as? String }
        let items: [[String: Any]] = ids.isEmpty ? [] : ((try? SQLite.shared.query(
            "SELECT * FROM stockin_items WHERE session_id IN (\(ids.map { _ in "?" }.joined(separator: ",")))",
            ids.map { SQLite.Value.text($0) }
        )) ?? [])
        return ["sessions": sessions, "items": items]
    }
}

File: src/main/stockin.ts
import { BrowserWindow } from 'electron';
import { getDb, getSetting } from './db';
import { deviceId, deviceName } from './device';
import { findByCode } from './products';
import * as live from './live';
import { StockinSession, StockinItem, StockinPlanRow } from '../shared/types';

/**
 * Naskladnění — naskladnění zboží.
 *
 * Zboží dorazí v krabici, někdo ho projde kus po kuse a v e-shopu musí
 * přibýt na skladě. Dělá se to u regálu, ne u počítače, takže se to začíná
 * na telefonu a dokončuje na Macu — a mezi tím se nesmí nic ztratit.
 *
 * Naskladnění je proto **seznam řádků, ne jeden dokument**: každý řádek je kód
 * a počet. Řádky se mezi zařízeními slučují po jednom (stejně jako deníky
 * poukazů), takže když se na telefonu přidá pět položek a na Macu tři,
 * výsledek má osm — bez ohledu na to, které zařízení bylo první.
 *
 * Co se **nedělá**: nikde se nezapisuje zásoba naslepo. U každého řádku se
 * pamatuje, kolik toho bylo skladem v okamžiku načtení, takže při odeslání
 * je vidět „bylo 4, přidáváme 10" — a když se mezitím prodalo, je to poznat.
 */

function now(): string {
  return new Date().toISOString();
}

function touch(id: string): void {
  getDb().prepare('UPDATE stockin SET updated_at = ? WHERE id = ?').run(now(), id);
  pushSoon(id);
}

/* ---------- rychlý posel ---------- */

/**
 * Jedno naskladnění tak, jak ho druhá strana umí sloučit.
 *
 * Je to tentýž tvar, jaký nese soubor ve sdílené složce, jen o jedné
 * položce — `mergeStockin` na druhé straně tedy nepotřebuje nic nového a
 * ostatních naskladnění se to nedotkne.
 */
export function sessionSlice(id: string): { sessions: any[]; items: any[] } | null {
  const d = getDb();
  const session = d.prepare('SELECT * FROM stockin WHERE id = ?').get(id) as any;
  if (!session) return null;
  const items = d.prepare('SELECT * FROM stockin_items WHERE session_id = ?').all(id) as any[];
  return { sessions: [session], items };
}

const pushTimers = new Map<string, NodeJS.Timeout>();

/**
 * Poslat změnu druhému zařízení co nevidět.
 *
 * Odklad je schválně: při skenování jde pípnutí za pípnutím a bez něj by
 * se posílalo desetkrát za vteřinu. Necelá půlvteřina z toho udělá jednu
 * zprávu a je to pořád „hned".
 *
 * Když posel nedoručí (spojení není, druhá strana neposlouchá), nic se
 * neděje — sdílená složka to donese v běžném kole.
 */
export function pushSoon(id: string): void {
  const running = pushTimers.get(id);
  if (running) clearTimeout(running);
  pushTimers.set(id, setTimeout(() => {
    pushTimers.delete(id);
    const slice = sessionSlice(id);
    if (slice) live.publish('stockin', slice);
  }, 400));
}

export function listSessions(): StockinSession[] {
  const rows = getDb().prepare(
    `SELECT s.*, (SELECT COUNT(*) FROM stockin_items i WHERE i.session_id = s.id) AS lines,
            (SELECT COALESCE(SUM(qty), 0) FROM stockin_items i WHERE i.session_id = s.id) AS pieces
     FROM stockin s WHERE s.state != 'deleted'
     ORDER BY (s.state = 'open') DESC, s.updated_at DESC LIMIT 60`
  ).all() as any[];
  return rows.map(mapSession);
}

function mapSession(r: any): StockinSession {
  return {
    id: r.id,
    title: r.title ?? '',
    note: r.note ?? '',
    device: r.device ?? '',
    state: r.state === 'sent' ? 'sent' : 'open',
    createdAt: r.created_at ?? '',
    updatedAt: r.updated_at ?? '',
    sentAt: r.sent_at || null,
    lines: r.lines ?? 0,
    pieces: r.pieces ?? 0
  };
}

export function createSession(title = ''): StockinSession {
  const id = `${deviceId().slice(0, 6)}-${Date.now().toString(36)}`;
  const at = now();
  getDb().prepare(
    `INSERT INTO stockin (id, title, device, state, created_at, updated_at)
     VALUES (?,?,?,'open',?,?)`
  ).run(id, title || `Naskladnění ${new Date().toLocaleDateString('cs-CZ')}`, deviceName(), at, at);
  // Ať se druhé zařízení dozví o zahájení, ne až o prvním pípnutí
  pushSoon(id);
  return sessionOf(id)!;
}

export function sessionOf(id: string): StockinSession | null {
  const row = getDb().prepare(
    `SELECT s.*, (SELECT COUNT(*) FROM stockin_items i WHERE i.session_id = s.id) AS lines,
            (SELECT COALESCE(SUM(qty), 0) FROM stockin_items i WHERE i.session_id = s.id) AS pieces
     FROM stockin s WHERE s.id = ? AND s.state != 'deleted'`
  ).get(id) as any;
  return row ? mapSession(row) : null;
}

export function itemsOf(id: string): StockinItem[] {
  const rows = getDb().prepare(
    'SELECT * FROM stockin_items WHERE session_id = ? ORDER BY added_at DESC'
  ).all(id) as any[];
  return rows.map(r => ({
    code: r.code,
    productCode: r.product_code ?? '',
    title: r.title ?? '',
    label: r.label ?? '',
    qty: r.qty ?? 0,
    stockBefore: r.stock_before ?? null,
    addedAt: r.added_at ?? ''
  }));
}

/**
 * Přidá načtený kód. Když už na naskladnění je, jen se přičte počet.
 *
 * Vrací i to, co se našlo — u regálu je potřeba na první pohled vidět, že
 * pípnutí sedlo na správné zboží. Když se nenajde nic, **nic se nepřidá**:
 * naskladnění s řádkem „neznámý kód" by se nedala odeslat a při zápisu do
 * e-shopu by se stejně musela řešit ručně.
 */
export function addScan(id: string, raw: string, qty = 1):
  { added: boolean; item?: StockinItem; unknown?: string } {
  const hit = findByCode(raw);
  if (!hit) return { added: false, unknown: (raw ?? '').trim() };

  const d = getDb();
  const exists = d.prepare('SELECT qty FROM stockin_items WHERE session_id = ? AND code = ?')
    .get(id, hit.code) as { qty: number } | undefined;

  if (exists) {
    d.prepare('UPDATE stockin_items SET qty = qty + ? WHERE session_id = ? AND code = ?')
      .run(Math.max(1, qty), id, hit.code);
  } else {
    d.prepare(
      `INSERT INTO stockin_items (session_id, code, product_code, title, label, qty, stock_before, added_at)
       VALUES (?,?,?,?,?,?,?,?)`
    ).run(id, hit.code, hit.productCode, hit.title, hit.label, Math.max(1, qty), hit.stock, now());
  }
  touch(id);
  const item = itemsOf(id).find(one => one.code === hit.code)!;
  return { added: true, item };
}

export function setQty(id: string, code: string, qty: number): void {
  const d = getDb();
  if (qty <= 0) d.prepare('DELETE FROM stockin_items WHERE session_id = ? AND code = ?').run(id, code);
  else d.prepare('UPDATE stockin_items SET qty = ? WHERE session_id = ? AND code = ?').run(Math.round(qty), id, code);
  touch(id);
}

export function renameSession(id: string, title: string, note = ''): void {
  getDb().prepare('UPDATE stockin SET title = ?, note = ?, updated_at = ? WHERE id = ?')
    .run(title, note, now(), id);
  pushSoon(id);
}

/**
 * Smazání se musí dozvědět i druhé zařízení.
 *
 * Řádek se proto nezahodí, jen se označí za smazaný a takový zůstane
 * ve sdíleném souboru. Kdyby se smazal doopravdy, druhé zařízení by ho
 * při nejbližší synchronizaci poslalo zpátky — přesně to se dělo: co se
 * smazalo na telefonu, do minuty se zase objevilo.
 *
 * Náhrobky se po dvou měsících uklidí; do té doby už se nemá co vracet.
 */
export function deleteSession(id: string): void {
  const d = getDb();
  d.prepare('DELETE FROM stockin_items WHERE session_id = ?').run(id);
  d.prepare("UPDATE stockin SET state = 'deleted', updated_at = ? WHERE id = ?").run(now(), id);
  pushSoon(id);
  purgeTombstones();
}

function purgeTombstones(): void {
  const cutoff = new Date(Date.now() - 60 * 86_400_000).toISOString();
  getDb().prepare("DELETE FROM stockin WHERE state = 'deleted' AND updated_at < ?").run(cutoff);
}

/** Označí naskladnění za odeslanou — už se do ní nepřidává a nesynchronizuje se zpět. */
export function markSent(id: string): void {
  getDb().prepare("UPDATE stockin SET state = 'sent', sent_at = ?, updated_at = ? WHERE id = ?")
    .run(now(), now(), id);
  pushSoon(id);
}

/**
 * Podklad pro zápis do e-shopu.
 *
 * K řádkům se dohledají vnitřní čísla z feedu (`product_id`, `variant_id`) —
 * bez nich se do Upgates zapsat nedá, formulář ani API kód produktu nepoužívá.
 * Co se dohledat nepodaří, se nezamlčí: vrátí se s prázdným `productId`
 * a rozhraní to ukáže jako řádek k ručnímu dořešení.
 */
export function planOf(id: string): StockinPlanRow[] {
  const d = getDb();
  const items = itemsOf(id);
  const out: StockinPlanRow[] = [];

  for (const item of items) {
    const variant = d.prepare('SELECT * FROM product_variants WHERE code = ?').get(item.code) as any;
    const product = d.prepare('SELECT * FROM products WHERE code = ?')
      .get(variant ? variant.product_code : item.code) as any;
    const stockNow = variant ? variant.stock : product?.stock ?? null;

    out.push({
      code: item.code,
      title: item.title || product?.title_cz || item.code,
      label: item.label,
      qty: item.qty,
      productId: product?.product_id ?? '',
      variantId: variant?.variant_id ?? '',
      stockNow: stockNow ?? null,
      stockBefore: item.stockBefore,
      /*
       * Zásoba se od načtení změnila — mezitím se prodalo nebo někdo
       * naskladnil ručně. Není to chyba, ale je to jediná informace, kvůli
       * které se má člověk před odesláním podívat.
       */
      moved: item.stockBefore !== null && stockNow !== null && item.stockBefore !== stockNow
    });
  }
  return out;
}

/* ---------- synchronizace mezi zařízeními ---------- */

/**
 * Sloučení naskladnění ze sdílené složky.
 *
 * **Vyhrává novější verze celého naskladnění, ne jednotlivé řádky.**
 *
 * Původně se slučovalo po řádcích a počet se bral jako vyšší z obou stran,
 * aby se opakovanou synchronizací nic nenafouklo. Jenže tím se nedalo počet
 * snížit ani řádek smazat: soubor z minulého kola hodnotu vždycky vrátil.
 * Projevilo se to i na jediném zařízení — opravený počet se do minuty vrátil
 * na původní, protože ho vzkřísil soubor, který to samé zařízení předtím
 * zapsalo.
 *
 * Naskladnění se dělá po sobě, ne najednou: začne se u regálu na telefonu
 * a dokončí na počítači. Novější strana proto přebírá i seznam řádků —
 * včetně toho, co z něj zmizelo.
 *
 * Dvě věci jsou konečné bez ohledu na čas: **smazané** a **zapsané**. Stačí,
 * aby to řeklo jedno zařízení; jednou zapsané zboží se nemá naskladnit
 * podruhé a smazané se nemá vracet.
 */
export function mergeStockin(remote: any): void {
  if (!remote || !Array.isArray(remote.sessions)) return;
  const d = getDb();

  const upsertSession = d.prepare(
    `INSERT INTO stockin (id, title, note, device, state, created_at, updated_at, sent_at)
     VALUES (@id, @title, @note, @device, @state, @created_at, @updated_at, @sent_at)
     ON CONFLICT(id) DO UPDATE SET
       title = CASE WHEN excluded.updated_at > stockin.updated_at THEN excluded.title ELSE stockin.title END,
       note = CASE WHEN excluded.updated_at > stockin.updated_at THEN excluded.note ELSE stockin.note END,
       state = CASE
         WHEN stockin.state = 'deleted' OR excluded.state = 'deleted' THEN 'deleted'
         WHEN excluded.state = 'sent' THEN 'sent'
         ELSE stockin.state END,
       sent_at = CASE WHEN stockin.sent_at = '' THEN excluded.sent_at ELSE stockin.sent_at END,
       updated_at = MAX(stockin.updated_at, excluded.updated_at)`
  );
  const insertItem = d.prepare(
    `INSERT OR REPLACE INTO stockin_items
       (session_id, code, product_code, title, label, qty, stock_before, added_at)
     VALUES (@session_id, @code, @product_code, @title, @label, @qty, @stock_before, @added_at)`
  );

  const run = d.transaction(() => {
    /*
     * Nejdřív se zjistí, čí verze je novější — porovnává se se stavem *před*
     * zápisem hlavičky, protože ta si po sloučení odnese pozdější čas z obou
     * stran a rozdíl by zmizel.
     */
    const takeover = new Set<string>();
    for (const s of remote.sessions) {
      if (!s?.id) continue;
      const id = String(s.id);
      const mine = d.prepare('SELECT updated_at, state FROM stockin WHERE id = ?').get(id) as any;
      const theirs = s.updated_at ?? '';
      if (mine?.state === 'deleted') continue;
      if (!mine || theirs > (mine.updated_at ?? '')) takeover.add(id);
    }

    for (const s of remote.sessions) {
      if (!s?.id) continue;
      upsertSession.run({
        id: String(s.id),
        title: s.title ?? '',
        note: s.note ?? '',
        device: s.device ?? '',
        state: s.state === 'sent' || s.state === 'deleted' ? s.state : 'open',
        created_at: s.created_at ?? now(),
        updated_at: s.updated_at ?? now(),
        sent_at: s.sent_at ?? ''
      });
    }

    // Přebírané naskladnění dostane cizí seznam řádků celý — a to znamená
    // i to, že řádky, které v něm nejsou, tady zmizí
    for (const id of takeover) {
      d.prepare('DELETE FROM stockin_items WHERE session_id = ?').run(id);
    }

    for (const i of remote.items ?? []) {
      if (!i?.session_id || !i?.code) continue;
      const id = String(i.session_id);
      if (!takeover.has(id)) continue;
      insertItem.run({
        session_id: id,
        code: String(i.code),
        product_code: i.product_code ?? '',
        title: i.title ?? '',
        label: i.label ?? '',
        qty: Number(i.qty) || 0,
        stock_before: i.stock_before ?? null,
        added_at: i.added_at ?? now()
      });
    }

    // Řádky smazaného naskladnění se nevracejí — ani z druhé strany
    d.prepare(
      "DELETE FROM stockin_items WHERE session_id IN (SELECT id FROM stockin WHERE state = 'deleted')"
    ).run();
  });
  run();
}

/** Co se má poslat do sdílené složky — jen rozpracované a nedávno odeslané. */
export function stockinExport(): { sessions: any[]; items: any[] } {
  const d = getDb();
  const cutoff = new Date(Date.now() - 30 * 86_400_000).toISOString();
  const sessions = d.prepare(
    "SELECT * FROM stockin WHERE state = 'open' OR updated_at > ? ORDER BY updated_at DESC LIMIT 120"
  ).all(cutoff) as any[];
  const ids = sessions.map(s => s.id);
  const items = ids.length
    ? d.prepare(
      `SELECT * FROM stockin_items WHERE session_id IN (${ids.map(() => '?').join(',')})`
    ).all(...ids) as any[]
    : [];
  return { sessions, items };
}

export function emitChanged(): void {
  for (const w of BrowserWindow.getAllWindows()) w.webContents.send('stockin:changed', {});
}

export { getSetting };
